tests/snapshot_tests.rs

```diff
@@ -0,0 +1,140 @@
+
+        let exitcode: i32 = fs::read_to_string(
+            format!("./tests/{test_dir}/exitcode"),
+        ).map(|s| {
+            s
+            .strip_suffix("\n")
+            .unwrap_or(s.as_str())
+            .parse::<i32>()
+            .unwrap_or(0)
+        }).unwrap_or(0);
+
+        let test_files = Self::_get_files(&test_dir);
+
+        SnapshotTest {
+            test_dir,
+            exitcode,
+            stderr,
+            stdout,
+            test_files,
+        }
+    }
+
+    #[cfg(not(feature = "save-snapshots"))]
+    fn execute(self) {
+        Command::cargo_bin(
+            env!("CARGO_PKG_NAME"),
+        )
+            .expect("binary to execute")
+            .args(self.test_files)
+            .assert()
+            .stdout(self.stdout)
+            .stderr(self.stderr)
+            .code(self.exitcode);
+    }
+
+    #[cfg(feature = "save-snapshots")]
+    fn execute(&self) {
+        let test_dir = self.test_dir.to_owned();
+        let test_files = self.test_files.to_owned();
+        let result = Command::cargo_bin(
+            env!("CARGO_PKG_NAME"),
+        )
+            .expect("binary to execute")
+            .args(test_files).ok().unwrap_or_else(|e| e.as_output().unwrap().to_owned());
+
+        if !result.stdout.is_empty() {
+            self._save_contents(
+                format!("./tests/{test_dir}/stdout"),
+                result.stdout,
+            );
+        }
+        if !result.stderr.is_empty() {
+            self._save_contents(
+                format!("./tests/{test_dir}/stderr"),
+                result.stderr,
+            );
+        }
+        if let Some(exitcode) = result.status.code() {
+            if exitcode > 0 {
+                self._save_contents(
+                    format!("./tests/{test_dir}/exitcode"),
+                    format!("{exitcode}\n").into(),
+                );
+            }
+        }
+    }
+
+    fn _save_contents(
+        &self,
+        file_name: String,
+        contents: Vec<u8>,
+    ) {
+        println!("Saving {}", file_name);
+        let res = File::create(file_name).unwrap().write_all(
+            &contents,
+        );
+        assert!(res.is_ok(), "{:?}", res);
+    }
+
+    fn _get_files(test_dir: &String) -> Vec<OsString> {
+        let yml = Some(OsStr::new("yml"));
+        fs::read_dir(
+            format!("./tests/{test_dir}"),
+        )
+        .unwrap()
+        .filter_map(Result::ok)
+        .filter(|f| f.path().extension() == yml)
+        .map(|f| f.path().into_os_string())
+        .collect::<Vec<OsString>>()
+    }
+}
+
+
+#[rstest]
+#[case("001_basic_workflow")]
+#[case("002_basic_action")]
+#[case("003_successful_globs")]
+#[case("004_failing_globs")]
+#[case("005_conditional_step_in_action")]
+#[case("006_workflow_dispatch_inputs_options")]
+#[case("007_funky_syntax")]
+#[case("008_job_dependencies")]
+#[case("009_multi_file")]
+#[cfg(not(feature = "remote-checks"))]
```

Comment:
I like this as it is, because I don't like it when a test suite makes any more assumptions about the local environment than it absolutely needs to (and "I have a real, working, fully-open Internet connection" is still, in some cases, an incorrect assumption).  I sometimes do dev without a working Internet connection, and a test suite that starts barfing red in all directions gives me the absolute pip.

In any event, I never trust developers (including myself!) to run all the tests before pushing (one cannot mandate that a pre-commit hook is in place in a remote repo, after all), so the most important thing is to make sure that CI is comprehensive -- which means, in this case, running the tests with `remote-checks` enabled.

It would be a nice "quality of life" thing if the test suite could print something at the end of a `cargo test` run saying "some tests weren't run because the `remote-checks` feature wasn't enabled", to give people a hint that re-running with extra features would be a good idea, but a lack of it wouldn't be a blocker for merge for me.

Reply:
@mpalmer when you put it like that I completely agree :D

Reply:
I see @mpalmer's point, however I made this decision for a different reason :) Interestingly, enabling `remote-checks` and running those first 9 tests resulted in different output for those tests! Hence why they were disabled for remote checks. We could alter those tests so they are idempotent to the issues they're testing, but I wanted to reduce the changes as much as I could.